### ISSUE TITLE
DataTemplateSelector.SelectTemplate is not called on iOS and Android

### ISSUE THREAD

**Issue by gritsenko** [unoplatform/uno#491]
## I'm submitting a...

 Bug report (I searched for similar issues and did not find one) 

Custom DataTemplateSelector is declared in ResourceDictionary and Bound as Static Resource to Button's content template selector

## Current behavior

Templates are not applied on Android and iOS builds _(and applied properly on UWP)_

**(overriden SelectTemplateCore methods are not called)**

## Expected behavior

Templates must be applied properly to each button

## Minimal reproduction of the problem with instructions

Project with reproduction of problem:
https://github.com/gritsenko/TemplateSelectorUno

## Environment

```
Nuget Package: 

Package Version(s): 

Affected platform(s):
- [x] iOS
- [x] Android
- [ ] WebAssembly **(not tested)**
- [ ] Windows
- [ ] Build tasks

Visual Studio
- [x] 2017 (version: )
- [ ] 2017 Preview (version: )
- [ ] for Mac (version: )

Relevant plugins
- [x] Resharper (version: )
```

**Comment by gritsenko** [comment]
there is also workaround to make it work:

```
   private void Button_OnLoaded(object sender, RoutedEventArgs e)
        {
            var btn = sender as Button;
            btn.ContentTemplate = btn.ContentTemplateSelector.SelectTemplate(btn.Content, btn);
        }

```

**Comment by carldebilly** [comment]
I investigated this bug report and I found 2 different problems: (one is not enough!)
1. **The current version of Uno is currently calling the wrong overload of `SelectTemplateCore()`**:
   Uno is calling the `SelectTemplateCore(object)` version and UWP is calling - _most of the time_ -
   the `SelectTemplateCore(object, DependencyObject)` version.
2. Many of the default _control templates_ defined in `generic.xaml` (in `Uno.UI`) are not binding
    the `ContentPresenter`'s `DataTemplateSelector` to their _templated parent_ respective property.

A simple workaround in your project would be to define both overloads (one calling the other) and
redefine the _Control Template_ for your problematic control (`Button` in your case).

I already did those fixes, but few tests remains before publishing it.
